### ISSUE TITLE
kqueue instead of epoll

### ISSUE THREAD

**Issue by colin-scott** [noxrepo/pox#15]
Mac OS X does not support epoll -- we may need to swap in kqueue under the hood for pox.lib.epoll_select if the OS is Mac.

For symptoms, run unit tests on a Mac.

**Comment by andiwundsam** [comment]
Epoll is supposed to be opt_in, with the regular select the default (and
portable) option.

Is epoll currently used by the unit tests?

 -A.

**Comment by colin-scott** [comment]
Yes, I believe so.

cs@fizzgig:~/Research/UCB/sdn-debugger/pox$ nosetests

# ..........EEE....

## ERROR: test_create (epoll_select_test.EpollSelectTest)

Traceback (most recent call last):
  File "/Users/cs/Research/UCB/sdn-debugger/pox/tests/unit/lib/epoll_select_test.py", line 44, in setUp
    self.es = EpollSelect()
  File "/Users/cs/Research/UCB/sdn-debugger/pox/pox/lib/epoll_select.py", line 9, in **init**
    self.epoll = select.epoll()
AttributeError: 'module' object has no attribute 'epoll'

# 

## ERROR: test_read_one_socket (epoll_select_test.EpollSelectTest)

Traceback (most recent call last):
  File "/Users/cs/Research/UCB/sdn-debugger/pox/tests/unit/lib/epoll_select_test.py", line 44, in setUp
    self.es = EpollSelect()
  File "/Users/cs/Research/UCB/sdn-debugger/pox/pox/lib/epoll_select.py", line 9, in **init**
    self.epoll = select.epoll()
AttributeError: 'module' object has no attribute 'epoll'

# 

## ERROR: test_write_more_sockets (epoll_select_test.EpollSelectTest)

Traceback (most recent call last):
  File "/Users/cs/Research/UCB/sdn-debugger/pox/tests/unit/lib/epoll_select_test.py", line 44, in setUp
    self.es = EpollSelect()
  File "/Users/cs/Research/UCB/sdn-debugger/pox/pox/lib/epoll_select.py", line 9, in **init**
    self.epoll = select.epoll()
AttributeError: 'module' object has no attribute 'epoll'

---

Ran 17 tests in 0.598s

**Comment by andiwundsam** [comment]
OK, I guess epoll_select_test should be a no-op unless on Linux.

 -- Andi

**Comment by colin-scott** [comment]
That would work temporarily, but I was actually thinking of a more permanent solution: have EpollSelect() choose either epoll or kqueue depending on the OS. This way we can keep the unit test there, and we can run experiments on Macs.

Anyway, not a big probem -- should certainly wait until after OSDI

**Comment by andiwundsam** [comment]
AFAIK the semantics of kqueue are quite a bit different from epoll.

I see select() as the normal / supported case that's portable, and
EpollSelect() as a Linux optimization for our scalability experiments.
/If/ we want to support >1024 Connections on Macs (the normal select
strategy will do fine for anything below that), we can add a parallel
KQueueSelectionStrategy.

And then have the Scheduler select one, based on OS and/or flags.

A.

**Comment by MurphyMc** [comment]
I think supporting epoll _really_ well may take changing semantics somewhat, and this may mean a separate recoco syscall and possibly a new IO loop written for epoll.  But at the moment, I don't think this is a major usecase, and the only real usecase for epoll is for simulation where it doesn't need to actually be phenomenal, it just needs to work with lots of sockets.

My points being:
1. The need for epoll/kqueue does not exist in most usecases
2. I think really, truly taking advantage of epoll (and probably kqueue) will take a bit of thought, work, and probably either a change in Select() semantics or an alternative syscall (there's no reason to EXPECT that you could take advantage of the benefits of epoll while using the interface of select, after all!)
3. I don't think 2 is currently necessary for any usecase.  It's a nice idea, but does anyone actually want to do it? (I'd be happy to, but according to my current thinking, it's going to be way, way down my list of priorities.)
4. _I_ am happy to leave epoll the way it is for Linux to support the simulator usecases
5. _I_ am not too worried about supporting the simulator usecases on Mac OS

I just wanted to chime in and see what you guys thought.  Should I be more worried about 5?  Have I missed any points?  Should we close the issue and resurrect it later if anyone is in danger of actually doing something about it or if the efficiency improvement of kqueue/epoll starts to look more necessary?  Or just leave it open?

**Comment by andiwundsam** [comment]
I agree, Murphy. I wrote the epoll implementation as a quick hack to do scalability experiments for the simulator under Linux. The code that makes the epoll loop expose a select-like interface to the recoco event loops is kind-of ugly and certainly not very runtime efficient. Making POX scale to 10k connections for real would require an improved overall IO architecture and a replacement of the "Select" recoco call with something more versatile (we would want to look at, e.g., how 
nginx does it for reference).

Anyway, I don't see many regular POX applications running into this, so I think we should punt on the issue for now. We've got bigger fish to fry :)
